### ISSUE TITLE
does libevent buffer split streams if there is more that one client connection?

### ISSUE THREAD

**Issue by hafizawang** [libevent/libevent#1079]
I'm trying to understand how bufferevent libevent on read works. I fully understand that streams might not come in one bufferevent on read (and could come in multiple) and i have cater for that by adding length as header of every transmitted packet as well as state checking on the server so that the server can distinguish individual packet to process. Everything works perfectly. But my design made some assumption which i cannot confirm. Perhaps someone here can clear this up. I hope i can spell my question clearly. so here it goes. currently im making an assumption like this. This question is for my server component.

server has connection to client a and b, each has its own bufferevent and each each is running on the same thread

client a sends packet containing 'abcdefghij'

client b sends packet containing 'klmnopqrst'

i'm making the assumption that not only streams for a packet can be received via multiple on read but also libevent could serve another client before finishing serving the previous client. I will attempt to illustrate my assumption. on the server side i assume this can happen.

from top to bottom (earlier to later timeline)

a bufferevent on read received 'abc'

b bufferevent on read received 'klmno'

a bufferevent on read received 'defgh'

b bufferevent on read received 'pqr'

a bufferevent on read received 'ij'

b bufferevent on read received 'st'

this is my current assumption and as such to cater for this each connection holds it's own 'personal' buffer and states and it works great. is this an overkill due to my wrong assumption or the scenario is more like below

a bufferevent on read received 'abc'

a bufferevent on read received 'defgh'

a bufferevent on read received 'ij'

b bufferevent on read received 'klmno'

b bufferevent on read received 'pqr'

b bufferevent on read received 'st'

in other words bufferevent on read serves only one connection until the packet stream is entirely received and then serves the other connection. I hope people can see what i'm asking here. if it's the second one i can avoid having each connection holds it's own 'personal' buffer and saves some memory space. thank you in advance.

**Comment by ploxiln** [comment]
The former case/assumption is true: read callbacks related to the two bufferevents can be interleaved. This is a major motivation of event-driven frameworks like libevent: you do not want "b" to be stuck if "a" pauses sending data.

**Comment by hafizawang** [comment]
Thank you very much. Glad to have that cleared up.
